### ISSUE TITLE
Overloading overlap leads to run incorrect method

### ISSUE THREAD

**Issue by arnvald** [crystal-lang/crystal#1047]
Hi!

I think the title of the issue might be quite confusing, but I'm not sure how to better describe it without code.

So I have following example:

```crystal
def tell_type(arg)
  "generic"
end

def tell_type(arg : (Int32|Float))
  "number"
end

def tell_type(arg : String)
  "string"
end

p "2 says #{tell_type(2)}"
p "\"2\" says #{tell_type("2")}"
p "3.2 says #{tell_type(3.2)}"
```

Running it of course returns:
```
"number"
"string"
"number"
```

However, when I change the 2nd method definition to:

```crystal
def tell_type(arg : (Int32|String)) # changed Float to String
  "number or string"
end
```

the result of code execution is following:
```
"2 says generic"
"\"2\" says string"
"3.2 says generic"
```

So the problem is that even though I have the special definition for `(Int32|String)`, generic method is executed.

Is it an expected behaviour or a bug in compiler?

**Comment by asterite** [comment]
The problem is that the definition for `String` is overwriting the definition for `Int32 | String`, which is a [bug](http://play.crystal-lang.org/#/r/8fw).

**Comment by asterite** [comment]
@arnvald Apart from the bug, is there a reason why you want to define an overload for `Int32 | String` and another for `String`?

**Comment by arnvald** [comment]
@asterite not really, I was just playing with type system in Crystal yesterday and I was wondering if I'd get some warning during compilation, saying that I defined method for `String` twice, and I accidentally found this bug.

**Comment by wmoxam** [comment]
This is fixed (not sure which version fixed it).

https://play.crystal-lang.org/#/r/11ph

**Comment by wmoxam** [comment]
Still a problem in 0.18  https://play.crystal-lang.org/#/r/11pn

**Comment by asterite** [comment]
This is fixed in 0.18.6: https://play.crystal-lang.org/#/r/12xi

```cr
def tell_type(arg)
  "generic"
end

def tell_type(arg : (Int32 | String)) # changed Float to String
  "number or string"
end

def tell_type(arg : String)
  "string"
end

puts "2 says #{tell_type(2)}"
puts "\"2\" says #{tell_type("2")}"
puts "3.2 says #{tell_type(3.2)}"
```

Output:

```
2 says number or string
"2" says string
3.2 says generic
```
